In [6]:
import keras
keras.__version__

'2.9.0'

<p dir="rtl" align="right">
<font size="+2" face="homa">
  <b>
مقدمه‌ای بر شبکه‌های کانولوشنی
  <br><br>
  <font size="+1" face="homa">
  فصل ۵ قسمت ۱ 
  </b>

<hr>

<p dir="rtl" align="justify">
<font face="homa">
در ابتدا، نگاه عملی به یک مثال شبکه کانولوشنی ساده خواهیم داشت. در این مثال از شبکه کانولوشنی برای دسته‌بندی ارقام MNIST استفاده می‌شود؛ کاری که در فصل 2 با شبکه تمام متصل (Dense) انجام دادیم (دقت آزمایش ما 8/97 درصد بود). با وجود ابتدایی بودن شبکه کانولوشنی، دقت آن چشم‌گیرتر از مدل تمام متصل -که در فصل 2 مشاهده کردیم- خواهد بود.
<br>
  <b>
مجموعه ‌داده MNIST
  </b>
<br>
کد زیر ساختار یک شبکه کانولوشنی پایه را به شما نشان می‌دهد. این شبکه پشته‌ای از لایه‌های conv2D و MaxPooling2D است..
<br>
موضوع مهم، شکل تنسورهای ورودی به شبکه کانولوشنی است (image_height, image_width, image_channels) (که حاوی بعد دسته نیست). در این مورد، پیکربندی شبکه کانولوشنی به صورتی خواهد بود که ورودی‌هایی را با اندازه (28,28,1) پردازش کند که اندازه تصاویر MNIST است.
 <br>

In [7]:
from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))

<p dir="rtl" align="justify">
<font face="homa">
حال می‌خواهیم معماری شبکه کانولوشنی را تا اینجای کار نشان دهیم:
<br><br>

In [8]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_3 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 13, 13, 32)       0         
 2D)                                                             
                                                                 
 conv2d_4 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 conv2d_5 (Conv2D)           (None, 3, 3, 64)          36928     
                                                                 
Total params: 55,744
Trainable params: 55,744
Non-trai

<p dir="rtl" align="justify">
<font face="homa">
می‌توان دید که خروجی هر لایه Conv2D و MAXPooling2D یک تنسور سه بعدی با شکل (height, width, channels) است. با پیشروی در عمق شبکه، ابعاد ارتفاع و عرض کوچک‌تر می‌شوند. تعداد کانال‌ها، اولین آرگومان لایه Conv2D است (32 یا 64).
<br>
مرحله بعد وارد کردن آخرین تنسور خروجی (با ابعاد (64, 3, 3)) به دسته‌بند تمام متصل است که قبلاً با آن آشنا شدید: پشته‌ای از لایه‌های تمام متصل. این دسته‌بند¬ها، ورودی خود را به صورت بردار تک بعدی است، دریافت می‌کنند، در حالی که خروجی فعلی تنسور سه بعدی است. ابتدا باید خروجی‌های سه بعدی را به شکل مسطح درآورده و یک بعدی کنیم و سپس چند لایه Dense در ادامه اضافه کنیم.

<br><br>

In [9]:
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

<p dir="rtl" align="justify">
<font face="homa">
دسته‌بندی با 10-خروجی را با استفاده از لایه نهایی با 10 خروجی و فعال‌سازی سافت مکس انجام خواهیم داد. حال شبکه به این شکل دیده می‌شود:
<br><br>

In [10]:
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d_3 (Conv2D)           (None, 26, 26, 32)        320       
                                                                 
 max_pooling2d_2 (MaxPooling  (None, 13, 13, 32)       0         
 2D)                                                             
                                                                 
 conv2d_4 (Conv2D)           (None, 11, 11, 64)        18496     
                                                                 
 max_pooling2d_3 (MaxPooling  (None, 5, 5, 64)         0         
 2D)                                                             
                                                                 
 conv2d_5 (Conv2D)           (None, 3, 3, 64)          36928     
                                                                 
 flatten (Flatten)           (None, 576)              

<p dir="rtl" align="justify">
<font face="homa">
همانطور که مشاهده می‌کنید، خروجی‌های (64, 3, 3) قبل از وارد شدن به دو لایه Dense به‌صورت بردارهایی با شکل (, 576) مسطح شده‌اند.
<br>
حال می‌خواهیم شبکه کانولوشنی را بروی ارقام MNIST آموزش دهیم. بسیاری از کدهای مثال MNIST در فصل 2 را مجدداً به کار خواهیم برد.
<br>

In [11]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

11490434/11490434 [==============================] - 0s 0us/step


In [12]:
model.compile(optimizer='rmsprop',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
model.fit(train_images, train_labels, epochs=5, batch_size=64)

Epoch 1/5
938/938 [==============================] - 61s 64ms/step - loss: 0.1689 - accuracy: 0.9467
Epoch 2/5
938/938 [==============================] - 54s 57ms/step - loss: 0.0483 - accuracy: 0.9846
Epoch 3/5
938/938 [==============================] - 55s 59ms/step - loss: 0.0336 - accuracy: 0.9896
Epoch 4/5
938/938 [==============================] - 55s 58ms/step - loss: 0.0242 - accuracy: 0.9922
Epoch 5/5
938/938 [==============================] - 53s 56ms/step - loss: 0.0200 - accuracy: 0.9940


<p dir="rtl" align="justify">
<font face="homa">
برای ارزیابی مدل روی نمونه‌های آزمایش:
<br><br>

In [13]:
test_loss, test_acc = model.evaluate(test_images, test_labels)

313/313 [==============================] - 4s 11ms/step - loss: 0.0302 - accuracy: 0.9903


In [14]:
test_acc

0.9902999997138977

<p dir="rtl" align="justify">
<font face="homa">
در حالی که دقت روی مجموعه داده آزمایش شبکه تمام متصل در فصل 2 برابر با 97/8 درصد بود، دقت شبکه کانولوشنی اولیه برابر با 99/3 درصد است: نرخ خطا را تا 68 درصد (نسبی) پایین آوردیم. بد نیست!
<br>
